Fix parsing history timeline file and archive recovery

red1452 opened the pull request:
Don't read past current TLI during archive recovery

If you set the recovery target timeline, the recovery would not
correctly follow the timeline history to reach that timeline, if a
local WAL segment had more WAL records on one of the older
timelines. Fix that by not reading WAL past the switchpoint in the
timeline history.

This case came up with pg_rewind: if you call pg_rewind on a server
that is a direct ancestor of the source, it exits with "no rewind
required" and does nothing. However, if the target had already
streamed WAL past the point of divergence, and just hadn't applied it
yet, when you start it up it will follow the original timeline, not
the one you tried to rewind to. This commit alone doesn't fix that,
because the server still won't know that it's supposed to follow the
other timeline. But this at least fixes the similar case when you have
set the recovery target timeline explicitly.

Discussion: https://www.postgresql.org/message-id/18575-1b9b3b60e2a480de%40postgresql.org

Original patch has been simply modified to correctly apply to our fork:
1) File xlogrecovery.c was renamed to xlog.c
2) Variable flushedUpto was renamed to receivedUpto
3) Callback wait_for_replay_catchup($primary2) was changed to
wait_for_catchup($primary2, 'replay', $primary1->lsn('flush'))

---

 Add test for check replication after recovery
    
 This commit adds new test to check correctness of standby's recovery. Standby
 should correctly connect to the primary's replication slot after recovery.

---
This PR should be rebased to preserve authorship.

hlinnaka committed:
Don't read past current TLI during archive recovery (#1189)

If you set the recovery target timeline, the recovery would not
correctly follow the timeline history to reach that timeline, if a
local WAL segment had more WAL records on one of the older
timelines. Fix that by not reading WAL past the switchpoint in the
timeline history.

This case came up with pg_rewind: if you call pg_rewind on a server
that is a direct ancestor of the source, it exits with "no rewind
required" and does nothing. However, if the target had already
streamed WAL past the point of divergence, and just hadn't applied it
yet, when you start it up it will follow the original timeline, not
the one you tried to rewind to. This commit alone doesn't fix that,
because the server still won't know that it's supposed to follow the
other timeline. But this at least fixes the similar case when you have
set the recovery target timeline explicitly.

Discussion: https://www.postgresql.org/message-id/18575-1b9b3b60e2a480de%40postgresql.org

Original patch has been simply modified to correctly apply to our fork:
1) File xlogrecovery.c was renamed to xlog.c
2) Variable flushedUpto was renamed to receivedUpto
3) Callback wait_for_replay_catchup($primary2) was changed to
wait_for_catchup($primary2, 'replay', $primary1->lsn('flush'))

RekGRpth commented:
Maybe a notice should be added to the PR description so that they don't squash it to preserve authorship?
And also add to the commit description what was changed compared to the original patch?

whitehawk commented on src/test/recovery/t/044_change_recovery_target.pl:
```diff
@@ -0,0 +1,98 @@
+#
+# Standby has replayed the WALL on TLI up to point S, and is currently
+# stopped. However, it already has all the WAL from TLI 1 locally in
+# pg_wal, because it was restored from th archive or streamed from the
```
'from th archive' -> 'from the archive'

red1452 replied:
fixed.

whitehawk commented on src/test/recovery/t/044_change_recovery_target.pl:
```diff
@@ -0,0 +1,98 @@
+$standby->start;
+
+# Pause the standby
+$standby->safe_psql('postgres', "SELECT pg_wal_replay_pause()");
```
Interesting thing to note:
in latest Postgres (at least 14 and later) following is said in the [documentation](https://www.postgresql.org/docs/14/functions-admin.html) for `pg_wal_replay_pause ()`:
> Request to pause recovery. A request doesn't mean that recovery stops right away. If you want a guarantee that recovery is actually paused, you need to check for the recovery pause state returned by pg_get_wal_replay_pause_state().

In Postgres 12 `pg_wal_replay_pause ()` "pauses recovery immediately". So current code is Ok for GPDB 7, but on the next major version, once we bump Postgres, we may get this test flaky. Maybe at least add a comment above the invocation of  `pg_wal_replay_pause()` that we'll need to add a check for `pg_get_wal_replay_pause_state()` later?

red1452 replied:
Added.

RekGRpth commented:
Fix regression tests, please

red1452 commented:
> Fix regression tests, please

Fixed.

whitehawk commented on src/test/isolation2/sql/fts_doublefault.sql:
```diff
@@ -122,53 +120,5 @@ select gp_request_fts_probe_scan();
 
 -1U: drop table last_timestamp;
 
--- Test case when mirror's promotion is interrupted after history file is created
```
Why is the test case removed?

red1452 replied:
Sorry, these changes should not be here.

red1452 replied:
fixed.

whitehawk commented on src/backend/access/transam/timeline.c:
```diff
@@ -78,7 +78,7 @@ readTimeLineHistory(TimeLineID targetTLI)
 	char		histfname[MAXFNAMELEN];
 	FILE	   *fd;
-	TimeLineHistoryEntry *entry;
+	TimeLineHistoryEntry *entry, *prevEntry = NULL;
```
Lets split into 2 lines, so we have 1 variable declaration per line.

red1452 replied:
Done.

RekGRpth commented on src/backend/access/transam/timeline.c:
```diff
@@ -174,11 +175,31 @@ readTimeLineHistory(TimeLineID targetTLI)
+		 */
+		if (prevEntry && prevend > end)
+		{
+			prevEntry->tli = tli;
+			prevEntry->end = end;
```
1) How are these lines used?
2) Why do we need the new variable `prevEntry` at all?
3) Why is the condition on the variable `prevend` not enough?

red1452 replied:
If we found, that new timeline starts from pointer, which is less, than end pointer of previous timeline, we should remove previous timeline. At the code I don't remove it, I just change previous timeline to current. Current timeline starts from pointer, which is the same as at removed timeline.

RekGRpth replied:
I don't see from the code that you are using these `tli` and `end` values ​​stored in `prevEntry` anywhere.

red1452 replied:
`prevEntry` is stored to get access at the next iteration in "for loop". I do not use field from `prevEntry`, I just change it. Because I can check such case only at the next iteration, when new line will be read from the history file.

RekGRpth replied:
> I do not use field from `prevEntry`, I just change it.

Why?

red1452 replied:
When current timeline is stored to the list, I save point to this timeline. At the next iteration I check new pointer (which is start pointer for the next timeline and end pointer for current timeline) if previous end pointer is bigger, I change field at `prevEntry`.

RekGRpth replied:
What will change if you remove these two lines?

red1452 replied:
for example: you have such history file:
```
1	0/C000090	at restore point "rp_repl_bug"
2	0/14000090	at restore point "rp_repl_bug2"
3	0/C000090	at restore point "rp_repl_bug"
```
The main idea of the history file is the next timeline starts from the end of the previous timeline. Here:
Timeline 1: 0/0000000 - 0/C000090
Timeline 2: 0/C000090 - 0/14000090
Timeline 3: 0/14000090 - 0/C000090 - incorrect
Timeline 4: 0/C000090 - ...

When we parse line 2, we create `TimeLineHistoryEntry` and fills fields: tli - `2`, start `0/C000090`, end `0/14000090`. At this moment it looks correctly. We also save current `TimeLineHistoryEntry` pointer to the `prevEntry` and go to the next iteration.
We parse line 3 and get: tli - `3`, start `0/14000090`, end `0/C000090`. Now, we can check start and end pointers. It is incorrect, because end pointer is less than start.
My patch gets `prevEntry` (where tli 2 was saved) and changes it: tli - `3`, start `0/C000090` (it was not changed), end `0/C000090`.
I save `prevEntry` not to find it in the list when we find such cases.

RekGRpth replied:
How about such simplification?
```diff
diff --git a/src/backend/access/transam/timeline.c b/src/backend/access/transam/timeline.c
index e6574013bdc..2b64a1db7ab 100644
--- a/src/backend/access/transam/timeline.c
+++ b/src/backend/access/transam/timeline.c
@@ -79,7 +79,6 @@ readTimeLineHistory(TimeLineID targetTLI)
 	char		histfname[MAXFNAMELEN];
 	FILE	   *fd;
 	TimeLineHistoryEntry *entry;
-	TimeLineHistoryEntry *prevEntry = NULL;
 	TimeLineID	lasttli = 0;
 	XLogRecPtr	prevend;
 	bool		fromArchive = false;
@@ -185,12 +184,10 @@ readTimeLineHistory(TimeLineID targetTLI)
 		 * previous timeline should be ignored, because the current timeline was
 		 * promoted from another timeline (not previous).
 		 */
-		if (prevEntry && prevend > end)
+		if (prevend > end)
 		{
-			prevEntry->tli = tli;
-			prevEntry->end = end;
-			prevend = end;
-			continue;
+			list_delete(result, entry);
+			pfree(entry);
 		}
 
 		entry = (TimeLineHistoryEntry *) palloc(sizeof(TimeLineHistoryEntry));
@@ -198,7 +195,6 @@ readTimeLineHistory(TimeLineID targetTLI)
 		entry->begin = prevend;
 		entry->end = end;
 		prevend = entry->end;
-		prevEntry = entry;
 
 		/* Build list with newest item first */
 		result = lcons(entry, result);
```

red1452 replied:
Your variant is less optimal, because it adds extra `pfree` and `palloc`, which can make code slower. And You just removed timeline 2 (if you take the example), but you should get start pointer from timeline 2 and end pointer from timeline 3 to get correct pointers for timeline 3.

RekGRpth replied:
ok, how about such simplification?
```diff
diff --git a/src/backend/access/transam/timeline.c b/src/backend/access/transam/timeline.c
index e6574013bdc..b3235ca68f1 100644
--- a/src/backend/access/transam/timeline.c
+++ b/src/backend/access/transam/timeline.c
@@ -78,8 +78,7 @@ readTimeLineHistory(TimeLineID targetTLI)
 	char		path[MAXPGPATH];
 	char		histfname[MAXFNAMELEN];
 	FILE	   *fd;
-	TimeLineHistoryEntry *entry;
-	TimeLineHistoryEntry *prevEntry = NULL;
+	TimeLineHistoryEntry *entry = NULL;
 	TimeLineID	lasttli = 0;
 	XLogRecPtr	prevend;
 	bool		fromArchive = false;
@@ -185,10 +184,10 @@ readTimeLineHistory(TimeLineID targetTLI)
 		 * previous timeline should be ignored, because the current timeline was
 		 * promoted from another timeline (not previous).
 		 */
-		if (prevEntry && prevend > end)
+		if (entry && prevend > end)
 		{
-			prevEntry->tli = tli;
-			prevEntry->end = end;
+			entry->tli = tli;
+			entry->end = end;
 			prevend = end;
 			continue;
 		}
@@ -198,7 +197,6 @@ readTimeLineHistory(TimeLineID targetTLI)
 		entry->begin = prevend;
 		entry->end = end;
 		prevend = entry->end;
-		prevEntry = entry;
 
 		/* Build list with newest item first */
 		result = lcons(entry, result);
```

red1452 replied:
Done.

RekGRpth commented on src/backend/access/transam/timeline.c:
```diff
@@ -174,11 +175,31 @@ readTimeLineHistory(TimeLineID targetTLI)
+		end = ((uint64) (switchpoint_hi)) << 32 | (uint64) switchpoint_lo;
+
+		/*
+		 * In case of some recovery we can get such history:
+		 * 1 -> 2
+		 *      2 -> 3
+		 * 1 -> 4
+		 *
+		 * Start pointer for timeline 4 is less than start pointer for timeline 3.
+		 * We can remove timeline 3 and get such history: 1 -> 2 -> 4.
```
I think that a comment on the code should describe the general approach, rather than show a specific example.

red1452 replied:
Changed.

whitehawk commented on src/backend/access/transam/timeline.c:
```diff
@@ -174,11 +175,31 @@ readTimeLineHistory(TimeLineID targetTLI)
+		 * 1 -> 4
+		 *
+		 * Start pointer for timeline 4 is less than start pointer for timeline 3.
+		 * We can remove timeline 3 and get such history: 1 -> 2 -> 4.
```
Why do we handle such case during reading of timeline history file? Maybe better to create already cleaned up history file, with removed not needed timelines?

red1452 replied:
It is good point. But history file is gotten from archive during recovery and there is not guarantee, that it will be correct, because it may be generated by old version of GPDB.

RekGRpth commented on src/backend/access/transam/timeline.c:
```diff
@@ -174,12 +175,28 @@ readTimeLineHistory(TimeLineID targetTLI)
-		entry->end = ((uint64) (switchpoint_hi)) << 32 | (uint64) switchpoint_lo;
+		entry->end = end;
 		prevend = entry->end;
-
```
restore this line

red1452 replied:
Restored.

RekGRpth commented on src/backend/access/transam/timeline.c:
```diff
@@ -174,10 +175,27 @@ readTimeLineHistory(TimeLineID targetTLI)
+		 * previous timeline should be ignored, because the current timeline was
+		 * promoted from another timeline (not previous).
+		 */
+		if (entry && prevend > end)
+		{
```
how about
```suggestion
		if (prevend > end)
		{
			Assert(entry != NULL);
```
and remove `= NULL` change from function begin?

red1452 replied:
Done.

red1452 replied:
I reverted the last changes because of compilation error:
```
#4 78.66 timeline.c: In function ‘readTimeLineHistory’:
#4 78.66 timeline.c:192:36: error: ‘entry’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
#4 78.66   192 |                         entry->end = end;
#4 78.66       |                         ~~~~~~~~~~~^~~~~
#4 78.81 cc1: all warnings being treated as errors
#4 78.81 make[5]: *** [../../../../src/Makefile.global:1026: timeline.o] Error 1
#4 78.81 make[5]: *** Waiting for unfinished jobs....
#4 79.75 make[4]: *** [../../../src/backend/common.mk:39: transam-recursive] Error 2
#4 79.75 make[4]: *** Waiting for unfinished jobs....
#4 80.73 fe-secure-openssl.c:1042:19: warning: 'ENGINE_by_id' is deprecated [-Wdeprecat
```

RekGRpth replied:
hm...

RekGRpth commented:
> For example:
> 
>     1. At the timeline 1 we made backup.
> 
>     2. Restored these backup and got timeline 2.
> 
>     3. Made the second backup.
> 
>     4. Restored the second backup and got timeline 3.
> 
>     5. Restored the first backup and got timeline 4.
> 
> 
> Schema: 1 -> 2 2 -> 3 1 -> 4
> 
> At this case start pointer of timeline 4 may be less than start pointer of timeline 3. The history file will be parsed incorrectly. Start pointer of timeline 3 will be more than end pointer.
> 
> When we will try to create a replica from the first backup and configure replication from primary, replica will try to get WAL from pointer, which is bigger than actual pointer at the primary.
> 
> This patch adds check for such cases and remove timelines with incorrect values of start and end pointers. At the example we will get such result schema for replica: 1 -> 2 -> 4 (without patch it is incorrect: 1 -> 2 -> 3 -> 4).

Does the proposed algorithm support a more branched timeline tree?
```
001 -> 012
       012 -> 013
              013 -> 014
                     014 -> 015
              013 -> 024
                     024 -> 025
       012 -> 023
              023 -> 034
                     034 -> 035
              023 -> 044
                     044 -> 045
101 -> 112
       112 -> 113
              113 -> 114
                     114 -> 115
              113 -> 124
                     124 -> 125
       112 -> 123
              123 -> 134
                     134 -> 135
              123 -> 144
                     144 -> 145
201 -> 212
       212 -> 213
              213 -> 214
                     214 -> 215
              213 -> 224
                     224 -> 225
       212 -> 223
              223 -> 234
                     234 -> 235
              223 -> 244
                     244 -> 245
```

red1452 commented:
> > For example:
> > ```
> > 1. At the timeline 1 we made backup.
> > 
> > 2. Restored these backup and got timeline 2.
> > 
> > 3. Made the second backup.
> > 
> > 4. Restored the second backup and got timeline 3.
> > 
> > 5. Restored the first backup and got timeline 4.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Schema: 1 -> 2 2 -> 3 1 -> 4
> > At this case start pointer of timeline 4 may be less than start pointer of timeline 3. The history file will be parsed incorrectly. Start pointer of timeline 3 will be more than end pointer.
> > When we will try to create a replica from the first backup and configure replication from primary, replica will try to get WAL from pointer, which is bigger than actual pointer at the primary.
> > This patch adds check for such cases and remove timelines with incorrect values of start and end pointers. At the example we will get such result schema for replica: 1 -> 2 -> 4 (without patch it is incorrect: 1 -> 2 -> 3 -> 4).
> 
> Does the proposed algorithm support a more branched timeline tree?
> 
> ```
> 001 -> 012
>        012 -> 013
>               013 -> 014
>                      014 -> 015
>               013 -> 024
>                      024 -> 025
>        012 -> 023
>               023 -> 034
>                      034 -> 035
>               023 -> 044
>                      044 -> 045
> 101 -> 112
>        112 -> 113
>               113 -> 114
>                      114 -> 115
>               113 -> 124
>                      124 -> 125
>        112 -> 123
>               123 -> 134
>                      134 -> 135
>               123 -> 144
>                      144 -> 145
> 201 -> 212
>        212 -> 213
>               213 -> 214
>                      214 -> 215
>               213 -> 224
>                      224 -> 225
>        212 -> 223
>               223 -> 234
>                      234 -> 235
>               223 -> 244
>                      244 -> 245
> ```

At what times should I execute the mirror?

RekGRpth commented:
> > > For example:
> > > ```
> > > 1. At the timeline 1 we made backup.
> > > 
> > > 2. Restored these backup and got timeline 2.
> > > 
> > > 3. Made the second backup.
> > > 
> > > 4. Restored the second backup and got timeline 3.
> > > 
> > > 5. Restored the first backup and got timeline 4.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Schema: 1 -> 2 2 -> 3 1 -> 4
> > > At this case start pointer of timeline 4 may be less than start pointer of timeline 3. The history file will be parsed incorrectly. Start pointer of timeline 3 will be more than end pointer.
> > > When we will try to create a replica from the first backup and configure replication from primary, replica will try to get WAL from pointer, which is bigger than actual pointer at the primary.
> > > This patch adds check for such cases and remove timelines with incorrect values of start and end pointers. At the example we will get such result schema for replica: 1 -> 2 -> 4 (without patch it is incorrect: 1 -> 2 -> 3 -> 4).
> > 
> > 
> > Does the proposed algorithm support a more branched timeline tree?
> > ```
> > 001 -> 012
> >        012 -> 013
> >               013 -> 014
> >                      014 -> 015
> >               013 -> 024
> >                      024 -> 025
> >        012 -> 023
> >               023 -> 034
> >                      034 -> 035
> >               023 -> 044
> >                      044 -> 045
> > 101 -> 112
> >        112 -> 113
> >               113 -> 114
> >                      114 -> 115
> >               113 -> 124
> >                      124 -> 125
> >        112 -> 123
> >               123 -> 134
> >                      134 -> 135
> >               123 -> 144
> >                      144 -> 145
> > 201 -> 212
> >        212 -> 213
> >               213 -> 214
> >                      214 -> 215
> >               213 -> 224
> >                      224 -> 225
> >        212 -> 223
> >               223 -> 234
> >                      234 -> 235
> >               223 -> 244
> >                      244 -> 245
> > ```
> 
> At what times should I execute the mirror?

123 for example

red1452 commented:
> 001 -> 012
>        012 -> 013
>               013 -> 014
>                      014 -> 015
>               013 -> 024
>                      024 -> 025
>        012 -> 023
>               023 -> 034
>                      034 -> 035
>               023 -> 044
>                      044 -> 045
> 101 -> 112
>        112 -> 113
>               113 -> 114
>                      114 -> 115
>               113 -> 124
>                      124 -> 125
>        112 -> 123
>               123 -> 134
>                      134 -> 135
>               123 -> 144
>                      144 -> 145
> 201 -> 212
>        212 -> 213
>               213 -> 214
>                      214 -> 215
>               213 -> 224
>                      224 -> 225
>        212 -> 223
>               223 -> 234
>                      234 -> 235
>               223 -> 244
>                      244 -> 245

I created new ticket for this problem.

RekGRpth commented on src/backend/access/transam/timeline.c:
```diff
@@ -112,6 +112,7 @@ readTimeLineHistory(TimeLineID targetTLI)
 		entry = (TimeLineHistoryEntry *) palloc(sizeof(TimeLineHistoryEntry));
 		entry->tli = targetTLI;
 		entry->begin = entry->end = InvalidXLogRecPtr;
+
```
?

RekGRpth commented on src/backend/access/transam/timeline.c:
```diff
@@ -174,10 +176,12 @@ readTimeLineHistory(TimeLineID targetTLI)
 		lasttli = tli;
 
+		end = ((uint64) (switchpoint_hi)) << 32 | (uint64) switchpoint_lo;
+
 		entry = (TimeLineHistoryEntry *) palloc(sizeof(TimeLineHistoryEntry));
 		entry->tli = tli;
 		entry->begin = prevend;
-		entry->end = ((uint64) (switchpoint_hi)) << 32 | (uint64) switchpoint_lo;
+		entry->end = end;
```
I don't understand what the essence of the changes is?

red1452 replied:
I did not fully remove old changes, which are not actual now.

RekGRpth commented:
Can the second commit be considered separately without the first one?

red1452 commented:
> Can the second commit be considered separately without the first one?

All problems, which were found during research at ADBDEV-6322, are fixed by both of commits. Test, which is present in the second commit, will fail, if there is not the first commit.

red1452 commented:
> Does the proposed algorithm support a more branched timeline tree?

Your algorithm now can be checked. And the check is success.

RekGRpth commented:
> > Can the second commit be considered separately without the first one?
> 
> All problems, which were found during research at ADBDEV-6322, are fixed by both of commits. Test, which is present in the second commit, will fail, if there is not the first commit.

Isn't the problem in task 7125 solved by just one second commit?

red1452 commented:
> > > Can the second commit be considered separately without the first one?
> > 
> > 
> > All problems, which were found during research at ADBDEV-6322, are fixed by both of commits. Test, which is present in the second commit, will fail, if there is not the first commit.
> 
> Isn't the problem in task 7125 solved by just one second commit?

Yes, it is.

RekGRpth commented:
> > > > Can the second commit be considered separately without the first one?
> > > 
> > > 
> > > All problems, which were found during research at ADBDEV-6322, are fixed by both of commits. Test, which is present in the second commit, will fail, if there is not the first commit.
> > 
> > 
> > Isn't the problem in task 7125 solved by just one second commit?
> 
> Yes, it is.

Then I suggest making a separate PR for this with appropriate tests.

red1452 commented:
> Then I suggest making a separate PR for this with appropriate tests.

It may be done by ADBDEV-7125.

RekGRpth commented:
> > Then I suggest making a separate PR for this with appropriate tests.
> 
> It may be done by ADBDEV-7125.

So I propose to separate the second commit from this PR into a new PR for task 7125 and accept it first.

red1452 commented:
> So I propose to separate the second commit from this PR into a new PR for task 7125 and accept it first.

Moved to #1235

RekGRpth commented on src/bin/pg_basebackup/t/050_check_recovery_backup.pl:
```diff
@@ -81,6 +124,9 @@ sub wait_recovery_and_switch_wal_primary
 $primary->safe_psql('postgres', 'CREATE ROLE postgres WITH LOGIN REPLICATION');
 
+# Initialize replica
+$replica->init(extra => ['--data-checksums']);
```
why not `$replica->init_from_backup`?

RekGRpth commented on src/bin/pg_basebackup/t/050_check_recovery_backup.pl:
```diff
@@ -19,6 +19,7 @@
 my $archivedir =  $tempdir . '/archive';
 
 my $primary = PostgreSQL::Test::Cluster->new('primary');
+my $replica = PostgreSQL::Test::Cluster->new('replica');
```
use name `standby` instead of `replica`

red1452 committed:
Add test for check replication after recovery (#1189)

In case of non-linear order of backups we can get case, when history file
contains information about timelines, which were after timeline, when such
backup was created. It means, that WAL files may be read from the incorrect (for
the backup) timelines, and such backup may no be recovered.

This commit adds new test to check correctness of standby's recovery. Standby
should correctly connect to the primary's replication slot after recovery.

RekGRpth commented:
> Variable readUpto was renamed to receivedUpto

may be `flushedUpto` to `receivedUpto`?

RekGRpth commented on src/test/recovery/t/044_change_recovery_target.pl:
```diff
@@ -0,0 +1,101 @@
+# Pause the standby
+# After bump to PostgreSQL 14 or newest we should add check by executing
+# pg_get_wal_replay_pause_state(), because now there is guarantee, that recovery
+# is actually paused, but after bump to PostgreSQL 14 we should check state.
```
may be
```suggestion
# is actually paused, but before bump to PostgreSQL 14 we should check state.
```

red1452 replied:
I think, there is not error here, check is needed after bump to PostgreSQL 14.

RekGRpth commented on src/test/recovery/t/050_check_recovery_backup.pl:
```diff
@@ -0,0 +1,162 @@
+$primary->start;
+$primary->safe_psql('postgres', 'CREATE DATABASE test_db');
+$primary->safe_psql('test_db', 'CREATE TABLE test AS SELECT generate_series(1,10)');
+$primary->safe_psql('postgres', 'CREATE ROLE postgres WITH LOGIN replicaTION');
```
```suggestion
$primary->safe_psql('postgres', 'CREATE ROLE postgres WITH LOGIN REPLICATION');
```

RekGRpth commented on src/test/recovery/t/050_check_recovery_backup.pl:
```diff
@@ -0,0 +1,162 @@
+$primary->safe_psql('postgres', 'CREATE ROLE postgres WITH LOGIN replicaTION');
+
+# Initialize standby
+$standby->init(extra => ['--data-checksums']);
```
What is the point of this command? And in general, `standby` is usually not initialized like this, but with `init_from_backup`.

RekGRpth commented:
Am I correct in understanding that the original case that required a second test is fixed by correct configuration, not by editing the code?

RekGRpth commented:
> > Am I correct in understanding that the original case that required a second test is fixed by correct configuration, not by editing the code?
> 
> Yes, you understand correctly.

Could you please provide a more detailed description of this issue and its solution in the second commit description?

whitehawk commented on src/test/recovery/t/050_check_recovery_backup.pl:
```diff
@@ -0,0 +1,159 @@
+
+	$standby->set_recovery_mode;
+	open my $conf, '>', "$standby_pgdata/postgresql.auto.conf";
+	print $conf "synchronous_standby_names = '*'\n";
```
do we need this config in the test?

RekGRpth commented on src/test/recovery/t/050_check_recovery_backup.pl:
```diff
@@ -0,0 +1,158 @@
+$primary->safe_psql('test_db', 'CREATE TABLE test AS SELECT generate_series(1,10)');
+$primary->safe_psql('postgres', 'CREATE ROLE postgres WITH LOGIN REPLICATION');
+
+create_backup_primary($basebackupdir1, 'backup_label1');
+$primary->stop('smart');
+restore_backup_primary($basebackupdir1, "recovery_target_name = 'backup_label1'\nrecovery_target_timeline = 'current'", '001 -> 012');
+wait_recovery_and_switch_wal_primary('001 -> 012');
+
+create_backup_primary($basebackupdir2, 'backup_label2');
+$primary->stop('smart');
+restore_backup_primary($basebackupdir2, "recovery_target_name = 'backup_label2'\nrecovery_target_timeline = 'current'", '012 -> 013');
+wait_recovery_and_switch_wal_primary('012 -> 013');
+
+$primary->stop('smart');
+restore_backup_primary($basebackupdir2, "recovery_target_name = 'backup_label2'\nrecovery_target_timeline = 'current'", '012 -> 023');
+wait_recovery_and_switch_wal_primary('012 -> 023');
+
+$primary->stop('smart');
+restore_backup_primary($basebackupdir1, "recovery_target_name = 'backup_label1'\nrecovery_target_timeline = 'current'", '101 -> 112');
+wait_recovery_and_switch_wal_primary('101 -> 112');
+
+$primary->stop('smart');
+restore_backup_primary($basebackupdir2, "recovery_target_name = 'backup_label2'\nrecovery_target_timeline = 'current'", '112 -> 113');
+wait_recovery_and_switch_wal_primary('112 -> 113');
+
+$primary->stop('smart');
+restore_backup_primary($basebackupdir1, "recovery_target_name = 'backup_label1'\nrecovery_target_timeline = 'current'", '101 -> 114');
+
+# wait for end of recovery and create replication slot
+$primary->poll_query_until('test_db', "SELECT true")
+  or die "Timed out while waiting for creating replication slot after recovery";
+$primary->safe_psql('test_db', "SELECT pg_create_physical_replication_slot('internal_wal_replication_slot')");
```
What do these tests with `primary` check? They all pass successfully and without a patch! Only the test with `standby` below does not pass without a patch.

red1452 replied:
These steps are needed to generate non-linear history of timelines.

RekGRpth replied:
Could you please write a little more about this in the comments to the test?

red1452 replied:
Added.

RekGRpth commented on src/test/recovery/t/050_check_recovery_backup.pl:
```diff
@@ -0,0 +1,158 @@
+	unlink "$standby_pgdata/postgresql.auto.conf";
+	$standby->set_standby_mode;
+	open $conf, '>', "$standby_pgdata/postgresql.auto.conf";
+	print $conf "primary_conninfo = ";
+	print $conf "'user=postgres port=" . $primary_port;
+	print $conf " host=" . $primary_host;
+	print $conf " sslmode=prefer sslcompression=1 gssencmode=prefer";
+	print $conf " krbsrvname=postgres target_session_attrs=any";
+	print $conf " application_name=gp_walreceiver'\n";
+	print $conf "primary_slot_name = 'internal_wal_replication_slot'\n";
```
Why do we need all these settings if the tests pass successfully without them?

red1452 replied:
Standby should correctly connect to standby after recovery. Replication slot is used for connect.

RekGRpth replied:
Still, I think that only these parameters are sufficient for the test, the other parameters are unimportant.
```suggestion
	print $conf "primary_conninfo = 'port=$primary_port host=$primary_host'\n";
	print $conf "primary_slot_name = 'internal_wal_replication_slot'\n";
```

red1452 replied:
Removed.

RekGRpth commented on src/test/recovery/t/050_check_recovery_backup.pl:
```diff
@@ -0,0 +1,158 @@
+# Initialize primary
+$primary->init(allows_streaming => 1, extra => ['--data-checksums']);
+$primary->append_conf("postgresql.conf", "archive_mode = 'on'");
+$primary->append_conf("postgresql.conf", "archive_command = 'cp -rv %p $archivedir/'");
```
Try to use the same flags as above.
```suggestion
$primary->append_conf("postgresql.conf", "archive_command = 'cp -vr %p $archivedir/'");
```
What are you using the "r" flag for? It seems to be for recursive copying of directories, and the archiver works only with one WAL-file.

red1452 replied:
Flag "r" was removed.

whitehawk commented:
This PR will be rebased. So the [commit](https://github.com/arenadata/gpdb/pull/1189/commits/6cabad1fcb767db0236ca3b0cd749a6d35884679) will appear as is. We need either to ammend it to the preceeding one, or make a proper description.

red1452 commented:
> This PR will be rebased. So the [commit](https://github.com/arenadata/gpdb/pull/1189/commits/6cabad1fcb767db0236ca3b0cd749a6d35884679) will appear as is. We need either to ammend it to the preceeding one, or make a proper description.

I rebased the branch, not it has only two commits.
